### PR TITLE
Unit tests for certificate rotation

### LINKED ISSUE #656: Unit tests for certificate rotation
After #555, we need to write e2e tests for the same as the minimum time for certificate expiration is 10 minutes so this will be a slow test.

### PR DESCRIPTION
**Pull request by Madhur97**
**What this PR does / why we need it**:
This PR refactors the certificate rotation code to extract the logic into a separate method and adds unit tests for this method.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #656 

**Special notes for your reviewer**

### CODE REVIEW

File: agent/host_agent_test.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	restclient "k8s.io/client-go/rest"
```
Comment (shivi28) [inline agent/host_agent_test.go:11]: Can we move this import from line 11 to somewhere between line 19-34

File: agent/host_agent_test.go [hunk 2]
```diff
@@ -861,4 +862,80 @@ users:
...
...RGdnRVBBRENDQVFvQ2dnRUJBSjZvCnpZSnhVUnhnK2VHMkM4TzJOQXZ2OU9KZXIzL0lZTVBqcXh3cTNKMHZEbm02L3lHK0RQNzFxNWo5MzdlWmdMSncKZkI1YVJMMzFjdUw3N2RFUjZXc3UwVGRCRGN3MGJ6TWFEQ1lRT016TGdESnlrYmhlUmNtRkNDb1Z1MHVqaXlSbApyL3U2YllEZVcveWxmeTgxS1RJaWRJU0U1RUhoYjBlYWZyT0htbTY1aEVYSU5CbkF0WmJiUjZXaVk0NGt5SG1KCnh4bXBORGdHRHh4cVJQZ2ZCbDhrS2NkSDYvZSt1b3dPVlJtalY3a3ZsNzdtUi9uOEEvTUM5Ujc3enh0TmVCcXcKYzlmK3hmODRrMXlVT0xSZHM4M1JrNUFHZjB3Q2VjSmxMUmwwakNyNkJvSStZRm8xUWtMSzluQVJKZGtFS0xvbApaVFNkbHRodnBWd0MvTHdHeklzQ0F3RUFBYU5aTUZjd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0hRWURWUjBPQkJZRUZDUmh3bm1rcW9ncUZwS09ObmNUaFgvV2VvQTlNQlVHQTFVZEVRUU8KTUF5Q0NtdDFZbVZ5Ym1WMFpYTXdEUVlKS29aSWh2Y05BUUVMQlFBRGdnRUJBSnFWOHYxdC9kR0NrV0tXRzhxKwpWUTdqTHNZWkc5dUwyN2x6MVovYXFidUcxMURoemZCUHBxVDVSMHNkNXYwVFJUQ1lpK1JTVCs2UnB6Kzd1bytZCndWWUpBREwrcEE0UExTTGRGb1RUcHBFL1ZXcE5rbE0rZnhKa1U5cU5YUkRpN28xTzR4YWJOQUxRZWRwb3RWaVUKSFAxV3IxT3VNbDI0SW5vNkJYWU5CQ0JjT21mOWV3ZjV4Z2R5T2Z5eEh6WXI0dWJUb1dqT0hlaE1xUnQ1SUVqdApJRGQ2R0V0cG43VWVwdnFZNE50aU91UzFReUZodW13UGc2VmhtTjZZMFRrcjJrY2RYU3RtWVBxYzNGcU12RkRYClhDOTMzMkZrTjNxRWlNUFJvbUw0cDQvbHN2aUpLY0VpUG5iTTBCZzRWV1R0bjFqNWtwTVFNbC9TMU8rRkR6U20KWVlVPQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
```
Comment (shivi28) [inline agent/host_agent_test.go:883]: Can we use smaller string value for testing purpose?

File: agent/host_agent_test.go [hunk 3]
```diff
@@ -861,4 +862,80 @@ users:
...
+			bootstrapKubeConf *os.File
+			err               error
+		)
+		BeforeEach(func() {
+			bootstrapKubeConf, err = ioutil.TempFile("", "bootstrap-kubeconfig")
+			Expect(err).NotTo(HaveOccurred())
+			bootstrapKubeConfig = bootstrapKubeConf.Name()
+		})
+		AfterEach(func() {
+			Expect(os.Remove(bootstrapKubeConf.Name())).ShouldNot(HaveOccurred())
```
Comment (mayur-tolexo) [inline agent/host_agent_test.go:876]:
```suggestion
			Expect(os.Remove(bootstrapKubeConfig)).ShouldNot(HaveOccurred())
```

File: agent/host_agent_test.go [hunk 4]
```diff
@@ -861,4 +862,80 @@ users:
...
+		})
+		It("should return if certificate data is not valid", func() {
```
Comment (mayur-tolexo) [inline agent/host_agent_test.go:878]:
```suggestion
		It("should return error if certificate data is not valid", func() {
```

File: agent/host_agent_test.go [hunk 5]
```diff
@@ -861,4 +862,80 @@ users:
...
+    client-certificate-data: aW5WYWxpZERhdGE=
+    client-key-data: aW5WYWxpZERhdGE=
+`)
+			_, err = bootstrapKubeConf.Write(testbootstrapKubeconfigInvalid)
+			Expect(err).NotTo(HaveOccurred())
+			var config *restclient.Config
+			config, err = registration.LoadRESTClientConfig(bootstrapKubeConfig)
+			Expect(err).NotTo(HaveOccurred())
+			err = certRotation(klogr.New(), "test-host", config)
+			Expect(err).ShouldNot(HaveOccurred())
```
Comment (mayur-tolexo) [inline agent/host_agent_test.go:907]: I'm not sure what all errors can occur in `certRotation` function but do you need to verify the error returned is the same as expected?

File: agent/main.go [hunk 6]
```diff
@@ -246,34 +246,40 @@ func handleBootstrapFlow(logger logr.Logger, hostName string) error {
...
 
-		// Poll after every few seconds
-		time.Sleep(pollDuration)
+	totalTimeCert := cert.NotAfter.Sub(cert.NotBefore)
+
+	// if less than 20% time left, renew the certs.
+	// https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210222-kubelet-authentication.md#kubelet-authenticator-flow
+	if time.Now().After(cert.NotAfter.Add(totalTimeCert / -5)) {
+		logger.Info("certificate expiration time left is less than 20%, renewing")
+		if err = handleBootstrapFlow(logger, hostName); err != nil {
```
Comment (mayur-tolexo) [inline agent/main.go:276]: You are not returning error returned from `handleBootstrapFlow` function call.

Comment (mayur-tolexo) [reply agent/main.go:276]: ok, my bad. certRotation is called in an infinite loop after poll duration. This looks good. 👍

File: agent/host_agent_test.go (hunk 4 above)
Comment (Madhur97) [reply agent/host_agent_test.go:878]: Existing code does not return error if certificate data is not valid (it just logs the failures). So this test also checks that the method should return without error.

File: agent/main.go (hunk 6 above)
Comment (Madhur97) [reply agent/main.go:276]: This PR refactors the existing code. As per the existing code, the error from `handleBootstrapFlow` function call is logged and it is not returned to the caller. I am not sure but perhaps this was done to prevent the byoh agent from crashing when there is some error during certificate rotation.

File: agent/host_agent_test.go (hunk 5 above)
Comment (Madhur97) [reply agent/host_agent_test.go:907]: Existing code does not return error if certificate data is not valid (it just logs the failures). So this test also checks that the method should return without error.

File: agent/host_agent_test.go [hunk 7]
```diff
@@ -861,4 +862,80 @@ users:
 			Expect(err.Error()).Should(ContainSubstring("kubeconfig generation failed: hostname is not valid"))
 		})
 	})
+	Context("When the certRotation func is called", func() {
+		var (
+			bootstrapKubeConf *os.File
+			err               error
+		)
+		BeforeEach(func() {
+			bootstrapKubeConf, err = ioutil.TempFile("", "bootstrap-kubeconfig")
```
Comment (dharmjit) [inline agent/host_agent_test.go:871]: nit: s/bootstrapKubeConf/kubeConfig/g as certRotation will be called once the Bootstrap flow is completed.

File: agent/main.go [hunk 8]
```diff
@@ -246,34 +246,40 @@ func handleBootstrapFlow(logger logr.Logger, hostName string) error {
...
+		if err := certRotation(logger, hostName, config); err != nil {
 			return err
 		}
+		// Poll after every few seconds
+		time.Sleep(pollDuration)
+	}
+}
 
-		totalTimeCert := cert.NotAfter.Sub(cert.NotBefore)
+func certRotation(logger logr.Logger, hostName string, config *rest.Config) error {
```
Comment (dharmjit) [inline agent/main.go:257]: Suggestion: We have extracted all of the code except for the loop in this function. Could you see if it's okay to move the for loop [here](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/main/agent/main.go#L179)?

**Review by dharmjit — COMMENTED** [review]
some nits only.

File: agent/host_agent_test.go (hunk 1 above)
Comment (Madhur97) [reply agent/host_agent_test.go]: The restclient import was present in a different import group separated by a blank line, perhaps that's why goimports did not take care in this case.

File: agent/main.go (hunk 8 above)
Comment (Madhur97) [reply agent/main.go:257]: As the main method is already handling a lot of things, to avoid more complexity it would be better to keep the for loop in a method of its own.

**Review by dharmjit — APPROVED** [review]
LGTM! Thanks.

**Review by mayur-tolexo — APPROVED** [review]
LGTM Nice work 👍

**Comment by shivi28** [comment]
LGTM
